caput/pipeline.py

```diff
@@ -605,45 +643,164 @@ def run(self):
+                        if t is not None:
+                            self._task_head += ii
+                            break
                     continue
-                keys = str(task._out_keys)
-                msg = "%s produced output data product with keys %s."
-                msg = msg % (task.__class__.__name__, keys)
-                logger.debug(msg)
-                for receiving_task in self.tasks:
-                    receiving_task._pipeline_inspect_queue_product(task._out_keys, out)
+
+            if self.execution_order == "legacy":
+                # Advance the task pointer. If this is the last task
+                # in the list, it will get wrapped when selecting
+                # the next task
+                self._task_idx += 1
+
+            # Ensure the output(s) are correctly structured
+            out = self._check_task_output(out, task)
+
+            if out is None:
+                continue
+
+            # Queue outputs for any associated tasks
+            for key, product in zip(task._out_keys, out):
+                # Try to pass this product to each task
+                received = [
+                    recv._pipeline_queue_product(key, product)
+                    for recv in self.tasks
+                    if recv is not None
+                ]
+
+                if not any(received):
+                    # Just warn. This probably shouldn't happen, but there
+                    # could be some edge cases to deal with
+                    warnings.warn(
+                        f"Task {task!s} tried to pass key {key} "
+                        "but no task was found to accept it."
+                    )
+
+        # Pipeline is done
+        logger.info("FIN")
+
+    def _next_task(self):
+        """Get the next task to run from the task list.
+
+        Task is chosen based the following criteria:
+        - able to do something in its current state
+        - highest priority
+        - highest base priority
+        - next in pipeline config order
+
+        If no task is available to do anything, restart from the
+        pipeline task head.
+        """
+        # Get a list of tasks which are availble to run
+        available = []
+
+        for ii in range(len(self.tasks)):
+            # Loop through tasks starting at the current index. Including
+            # the current tasks first ensures we clear out completed
+            # tasks faster
+            jj = (ii + self._task_idx) % len(self.tasks)
+            task = self.tasks[jj]
+
+            if task is None:
+                continue
+
+            if task._pipeline_is_available:
+                available.append(jj)
+
+        if not available:
+            # Nothing is currently available, so fall back to a
+            # blind loop starting at the first task. If there is
+            # nothing left, this will raise StopIteration.
+            self._task_idx = self._task_head
+            return self._iter_tasks()
```

Comment:
Though I think it could still work, if you change what I suggest to put at the beginning of `_iter_tasks` above to
```
if self._task_idx is None:
    self._task_idx = self._task_head
else:
   self._task_idx += 1
```
and set `self._task_idx` to `None` here, though I don't know if that's really an improvement.

Reply:
Yeah, this is a tricky (and annoying) one, and I haven't really come up with a cleaner way of doing it. 

The thing here is that I **only** want the iterator to advance by one if I reach the point in the pipeline at line 707 *and* the method is `legacy`. In the exception handling above, I modify the task pointer and I want the next pipeline iteration to start at the new pointer location (inclusive). So, even if the execution order is `legacy`, I don't necessarily want to increment by one on every call to `_iter_tasks`.

Similarly when calling it as a fallback, I reset the task pointer to the head and I want it to include that task